### PR TITLE
Bring onDemand tablets online for live ingest, immediate scans

### PR DESCRIPTION
**Pull request by dlmarion**

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java [hunk 1]
```diff
@@ -211,9 +214,9 @@ public List<Short> check(Environment env, Mutation mutation) {
         } catch (NumberFormatException | ArrayIndexOutOfBoundsException nfe) {
           violations = addViolation(violations, 1);
         }
-      } else if (columnFamily.equals(ScanFileColumnFamily.NAME)) {
-
-      } else if (columnFamily.equals(BulkFileColumnFamily.NAME)) {
+      } else if (columnFamily.equals(ScanFileColumnFamily.NAME)) {} else if (columnFamily
```
Comment (ddanielr) [inline server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java:217]: Is there a future reason to have two separate `else if (x) {} else if (y) {}` no-op statements vs a single `else if (x or y)`? 

If so, there are now four distinct comparisons on `columnFamily.equals` for `TEXT` values. Could this be simplified into a switch statement for readability?

edit - `TEXT` isn't an enum subclass so ignore the switch statement comment.

Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java:217]: > Is there a future reason to have two separate else if (x) {} else if (y) {} no-op statements vs a single else if (x or y)?

No reason, just keeping my change in line with the others. We can certainly collapse them before this is merged.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java [hunk 2]
```diff
@@ -1555,4 +1556,19 @@ public TSummaries contiuneGetSummaries(TInfo tinfo, long sessionId)
       return handleTimeout(sessionId);
     }
   }
+
+  @Override
+  public void bringOnDemandTabletOnline(TInfo tinfo, TCredentials credentials,
+      TKeyExtent tkeyExtent) throws ThriftSecurityException, TException {
+    final TableId tableId = TableId.of(new String(tkeyExtent.getTable(), UTF_8));
+    NamespaceId namespaceId = getNamespaceId(credentials, tableId);
+    if (!security.canChangeTableState(credentials, tableId, FateOperation.TABLE_ONDEMAND,
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java]: If a table is in the ondemand state, then from a permission perspective as long as a user can read or write to a table then it seems like the user should be able to bring a tablet online.  For example if User A can change table state and User B can read the table, then after User A sets the table to on demand User B should be able to do immediate reads.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java [hunk 3]
```diff
@@ -524,6 +548,53 @@ public TabletLocation locateTablet(ClientContext context, Text row, boolean skip
...
+      if (scanRangeKE.overlaps(tabletExtent)) {
+        if (!tm.getOnDemand()) {
+          Location loc = tm.getLocation();
+          if (loc != null) {
+            log.debug("tablet {} has location of: {}:{}", loc.getType(), loc.getHostPort());
+          }
+          log.debug("Marking tablet as onDemand for extent: {}", tabletExtent);
+          ThriftClientTypes.TABLET_MGMT.executeVoid(context,
+              client -> client.bringOnDemandTabletOnline(TraceUtil.traceInfo(), context.rpcCreds(),
+                  tabletExtent.toThrift()));
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java]: Would be good to support passing a list of extents and avoid an RPC per extent.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java]: Fixed in e6c2939

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java (hunk 2 above)
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java]: Fixed in e6c2939

**Comment by dlmarion** [comment]
This has been merged into https://github.com/apache/accumulo/pull/3250

**Pull request closed without merging**